### PR TITLE
Allow custom layer URDF to use param:// URLs, add frame prefix setting

### PR DESCRIPTION
**Pull request by achim-k**
**User-Facing Changes**
- Custom layer URDF: Allow `param://` URLs and add frame prefix setting

**Description**
Adds the possibility to specify `param://<param_name>` URLs in the custom layer URDF. This will make it easier to visualize multiple URDFs when using datasources that provide parameter support (foxglove websocket, rosbridge, ros1 native). Besides allowing to load URDFs from a parameter, this PR also adds support for the frame prefix (also commonly called tfPrefix) setting. This setting allows to visualize different robots that share the same URDF definition but the [robot_state_publisher](https://index.ros.org/p/robot_state_publisher/github-ros-robot_state_publisher) publishes their transforms with a individual prefix.

Fixes FG-3292
Fixes FG-4057

Related: 
- https://github.com/orgs/foxglove/discussions/407
- https://github.com/orgs/foxglove/discussions/759
- https://github.com/orgs/foxglove/discussions/712

image

### CODE REVIEW

File: packages/studio-base/src/panels/ThreeDeeRender/renderables/Urdfs.ts [hunk 1]
```diff
@@ -247,6 +249,12 @@ export class Urdfs extends SceneExtension<UrdfRenderable> {
 
         const fields: SettingsTreeFields = {
           url: { label: "URL", input: "string", placeholder, help, value: config.url ?? "" },
+          framePrefix: {
+            label: "frame prefix",
```
Comment (snosenzo) [inline packages/studio-base/src/panels/ThreeDeeRender/renderables/Urdfs.ts:253]:
```suggestion
            label: "Frame prefix",
```

File: packages/studio-base/src/panels/ThreeDeeRender/renderables/Urdfs.ts [hunk 2]
```diff
@@ -456,10 +464,21 @@ export class Urdfs extends SceneExtension<UrdfRenderable> {
...
+    }
+
+    // Update custom layer URDFs that use param:// URLs.
+    for (const [instanceId, renderable] of this.renderables.entries()) {
+      const url = (renderable.userData.settings as Partial<LayerSettingsCustomUrdf>).url ?? "";
+      if (url.startsWith("param://")) {
+        const paramName = url.slice("param://".length);
+        const urdf = parameters?.get(paramName);
+        if (typeof urdf === "string") {
+          this.#loadUrdf(instanceId, urdf);
```
Comment (snosenzo) [inline packages/studio-base/src/panels/ThreeDeeRender/renderables/Urdfs.ts:478]: Do we need to do any cleanup of the old model when we load a new one? Are transforms cleaned up properly?

File: packages/studio-base/src/panels/ThreeDeeRender/renderables/Urdfs.ts [hunk 3]
```diff
@@ -668,13 +708,32 @@ export class Urdfs extends SceneExtension<UrdfRenderable> {
...
-async function parseUrdf(text: string): Promise<ParsedUrdf> {
+async function parseUrdf(text: string, framePrefix?: string): Promise<ParsedUrdf> {
   const fileFetcher = getFileFetch();
-
+  const applyFramePrefix = (name: string) => `${framePrefix}${name}`;
   try {
     log.debug(`Parsing ${text.length} byte URDF`);
     const robot = await parseRobot(text, fileFetcher);
 
+    if (framePrefix) {
```
Comment (snosenzo) [inline packages/studio-base/src/panels/ThreeDeeRender/renderables/Urdfs.ts:718]: why is it necessary to apply the frame prefix here?

**Review by snosenzo — APPROVED** [review]
not sure how easy it would be but it might be nice to get a story for this behavior as well.

File: packages/studio-base/src/panels/ThreeDeeRender/renderables/Urdfs.ts (hunk 2 above)
Comment (achim-k) [reply packages/studio-base/src/panels/ThreeDeeRender/renderables/Urdfs.ts:478]: You are right, that might be a problem. Especially when changing the frame prefix. Will look into this

**Comment by achim-k** [comment]
> not sure how easy it would be but it might be nice to get a story for this behavior as well.

Will look into this. Will also need to create a story for #6409, so can combine this effort

File: packages/studio-base/src/panels/ThreeDeeRender/renderables/Urdfs.ts (hunk 2 above)
Comment (achim-k) [reply packages/studio-base/src/panels/ThreeDeeRender/renderables/Urdfs.ts:473]: https://github.com/foxglove/studio/pull/6359/commits/d7b4e2eb3c8e8819c69670f6fda6690ea8a625ac is now cleaning up old transforms. However, the list of coordinate frames will grow, there seems to be no way to clean them up?

Comment (snosenzo) [reply packages/studio-base/src/panels/ThreeDeeRender/renderables/Urdfs.ts:473]: That must mean that they still have child frames with transforms in them. In `removeTransforms` I believe we clean up empty frames.

Comment (snosenzo) [reply packages/studio-base/src/panels/ThreeDeeRender/renderables/Urdfs.ts:473]: Maybe that's for the case where you've received transforms from the source under one prefix but change the prefix to another? I expect the frames received from the source would be there and also the frames under the new prefix. However I would expect changing the prefix again to clear the frames under the second prefix given it hasn't received transforms from the source.

Comment (achim-k) [reply packages/studio-base/src/panels/ThreeDeeRender/renderables/Urdfs.ts:473]: I just confirmed that the size of `renderer.transformTree` and `renderer.coordinateFrameList` remains the same when changing prefix. So it seems that old frames / transforms are cleaned up correctly.

**Review by snosenzo — APPROVED** [review]

**Pull request merged**